common/archival/client.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package archival
+
+import (
+	"context"
+	"errors"
+	"io"
+	"time"
+)
+
+// WriteStatus indicates if the write went through or not
+type WriteStatus int
+
+const (
+	// NoResponse indicates an unsuccessful write
+	NoResponse WriteStatus = iota
+	// Success indicates successful write, where something was written
+	Success
+	// NothingUpdated is a success and indicates that nothing needed to be updated
+	NothingUpdated
+)
+
+// ItemType indicates if the object is Directory or a Blob
+type ItemType int
+
+const (
+	// ItemTypeDirectory indicates that the object is a directory
+	ItemTypeDirectory ItemType = iota + 1
+
+	// ItemTypeBlob indicates that the object is a blob
+	ItemTypeBlob
+)
+
+// ListItem is a single listing in ListDirectoryResponse
+type ListItem struct {
+	Type      ItemType
+	Name      *string
+	CreatedAt *time.Time
+	Owner     *string
+	Size      *int64
+	Tags      map[string]string
+}
+
+// UploadBlobRequest is request for UploadBlob
+type UploadBlobRequest struct {
+	Path       *string
+	Body       io.Reader
+	Size       *int64
+	CreateDirs *bool
+	Tags       map[string]string
+	ChunkSize  *int64
+}
+
+// UploadBlobResponse is response from UploadBlob
+type UploadBlobResponse struct {
+	Status WriteStatus
+}
+
+// DownloadBlobRequest is request for DownloadBlob
+type DownloadBlobRequest struct {
+	Path        *string
+	Writer      io.WriterAt
+	IsStreaming *bool
```

Comment:
see comment above - for the caller, there should be no difference between streaming / non-streaming